tests/TaskLineRenderer.test.ts

```diff
@@ -0,0 +1,156 @@
+            return getDescriptionText(parentRender);
+        };
+
+        const descriptionWithFilter = await getDescriptionTest();
+        expect(descriptionWithFilter).toEqual('This is a simple task with a #global filter');
+
+        updateSettings({ globalFilter: '#global', removeGlobalFilter: true });
+        const descriptionWithoutFilter = await getDescriptionTest();
+        expect(descriptionWithoutFilter).toEqual('This is a simple task with a  filter');
+        resetSettings();
+    });
+
+    it('renders task components according to the given layout', async () => {
+        const testLayoutOptions = async (
+            taskLine: string,
+            layoutOptions: Partial<LayoutOptions>,
+            expectedRender: string,
+        ) => {
+            const task = fromLine({
+                line: taskLine,
+            });
+            const fullLayoutOptions = { ...new LayoutOptions(), ...layoutOptions };
+            const parentRender = await createMockParentAndRender(task, fullLayoutOptions);
+            const renderedDescription = getDescriptionText(parentRender);
+            expect(renderedDescription).toEqual(expectedRender);
+        };
+
+        const taskLine = '- [ ] Wobble ⏫ 📅 2022-07-02 ⏳ 2022-07-02 🛫 2022-07-02 🔁 every day';
+
+        // Test the default layout
+        await testLayoutOptions(taskLine, {}, 'Wobble ⏫ 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02');
+
+        // Without priority
+        await testLayoutOptions(
+            taskLine,
+            { hidePriority: true },
+            'Wobble 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02',
+        );
+
+        // Without start date
+        await testLayoutOptions(
+            taskLine,
+            { hideStartDate: true },
+            'Wobble ⏫ 🔁 every day ⏳ 2022-07-02 📅 2022-07-02',
+        );
+
+        // Without scheduled date
+        await testLayoutOptions(
+            taskLine,
+            { hideScheduledDate: true },
+            'Wobble ⏫ 🔁 every day 🛫 2022-07-02 📅 2022-07-02',
+        );
+
+        // Without due date
+        await testLayoutOptions(taskLine, { hideDueDate: true }, 'Wobble ⏫ 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02');
+
+        // Without recurrence rule
+        await testLayoutOptions(
+            taskLine,
+            { hideRecurrenceRule: true },
+            'Wobble ⏫ 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02',
+        );
+
+        const doneTask = '- [x] Wobble ✅ 2022-07-02 ⏫ 📅 2022-07-02 ⏳ 2022-07-02 🛫 2022-07-02 🔁 every day';
+
+        // Done task - default layout
+        await testLayoutOptions(
+            doneTask,
+            {},
+            'Wobble ⏫ 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02 ✅ 2022-07-02',
+        );
+
+        // Done task - without done date
+        await testLayoutOptions(
+            doneTask,
+            { hideDoneDate: true },
+            'Wobble ⏫ 🔁 every day 🛫 2022-07-02 ⏳ 2022-07-02 📅 2022-07-02',
+        );
```

Comment:
The comments here are useful in conveying the purpose of the test. I wondered if it would be possible to rework the code so that they could be moved to test names instead.

### Sharing `testLayoutOptions()` between `it` blocks

I found that if I moved `testLayoutOptions()` to just above this line...

```ts
    it('renders task components according to the given layout', async () => {
```

... then I could write individual test, each with a meaningful name.

For example I was then able to add this new test:

```ts
    it('should write placeholder message if a date is invalid', async () => {
        const taskLine = '- [ ] Task with invalid due date 📅 2023-13-02';
        await testLayoutOptions(taskLine, {}, 'Task with invalid due date 📅 Invalid date');
    });
```

I liked that a lot. I have been concerned for some time that there is no useful information in the `📅 Invalid date` text - but had no way of testing it.

### Easy TDD

I can imagine in future TDDing this to write a failing test like this, to show that I expect the invalid date to also be displayed:

```ts
    it('should write placeholder message with the date if a date is invalid', async () => {
        const taskLine = '- [ ] Task with invalid due date 📅 2023-13-02';
        await testLayoutOptions(taskLine, {}, 'Task with invalid due date 📅 Invalid date: 2023-13-02');
    });
```

### Alternative layout

I experimented with a different layout for the above tests - by inlining taskLine:

```ts
    it('should write placeholder message if a date is invalid', async () => {
        await testLayoutOptions(
            '- [ ] Task with invalid due date 📅 2023-13-02',
            {},
            'Task with invalid due date 📅 Invalid date',
        );
    });
```

The Task text being a bit closer to vertically aligned means I find it easier to quickly glance and see how the input differs from the output.

Another advantage of inlining the input data for the test is that the input is right next to the expected value.

In the current tests, the `taskLine` value is used in many tests, which means that:

- It has to contain all the input data for all the places it might be used, meaning the test purpose is less clear than if specifically useful data were chosen for each little test.
- If there is a test failure, the input is a long way from the failure
- It encourages writing long `it` blocks with a single description, which makes it harder to locate any failure in the source.

Reply:
Looking again at this code from my comment above... as well test name adding info, it helps when the task description itself can say more about the purpose.

```ts
    it('should write placeholder message if a date is invalid', async () => {
        const taskLine = '- [ ] Task with invalid due date 📅 2023-13-02';
        await testLayoutOptions(taskLine, {}, 'Task with invalid due date 📅 Invalid date');
    });
```

Reply:
Will do and restructure; I'm unsure if this is something you already did in your branch, so waiting to see the result of the merge.